### PR TITLE
Refresh unpromotable shards after primary write

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently, we have implemented a manual work around to ensure that
post-write refreshes are propagated to unpromotable shards. However,
this work should occur in the async after write action and ensure that
both forced refresh and wait_for refresh are implemented. This commit
resolves this by ensuring that the appropriate transport messages are
sent for a post-write refresh.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:80]: Could inline the function above since it's not called anywhere else.

File: server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:109]: Using ActionListener.completeWith() seems that it could remove having two `onResponse`s.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:415]: There can be more than one unpromotable shards, right?

If we want to fail the unpromotable shard(s), is there a reason we do not fail the shard in this PR? I am not sure yet how to recognize which shard failed. I wonder if the Exception has any field from which node it occurred. Else, we'd want to make `TransportBroadcastUnpromotableAction#doExecute()`'s response handler capture which node failed and pass it in the exception/response somehow.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:442]: Is there a corresponding decrease somewhere?

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:487]: What does this warning mean?

**Review by kingherc — COMMENTED** [review]
Looks into the right direction!
Would be nice to mention ES-5312 in PR&commit description.
Leaving some comments.

**Comment by pxsalehi** [comment]
Also I think we should we remove the work-around in this PR: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java#L197-L199

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:205]: Is it maybe because postWriteAction was just to simulate delays, and will be removed anyway soon?

File: server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java:132]: I wonder if we really need to make this change and the segment generation supplier at all?

My understanding (please correct me if I'm wrong) is that we don't need the refreshed segment generation on stateful Elasticsearch and we need it only for index shards on stateless. And in that case it is already provided through `InternalEngine#refresh(String)` when the write request's refresh policy is `IMMEDIATE` and when it is `WAIT_UNTIL` we only need to wait for the next refresh and then retrieve the latest commited segment generation which is already available with `getLastCommittedSegmentInfos`?

Happy to discuss this

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:1865]: That as a mistake

File: server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java:132]: This seems like a good idea. I have access to the `IndexShard` in post refresh action. And I can get that at minimum from the commit stats. I'll make this change tomorrow.

Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java:132]: I mean, I guess the biggest concern here is that this requires a flush happens for this refresh logic to work. Obviously this occurs in stateless, but all of the "unpromotable" logic really doesn't involve stateless at all. So I guess I would need a discoveryNode.isStateless check here.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:205]: Yes. Post write will leave soon.

Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:205]: Or be replaced by a dedicated durability action.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:415]: Yes. It will require multiple modifications and tests which is best left to a follow-up task.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:442]: Yes. `maybeFinish`

File: server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java:132]: We should maybe talk about this briefly tomorrow. Currently the refresh listener is called from the Lucene level. So I think that we would need to change the IndexEngine logic to ensure that a refresh call is ALWAYS forcing a flush. Right now we only do that from the api call. So without a change there, the periodic refresh could indicate that the doc is visible, but the generation still be the old generation (I think).

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:2025]: This is a possibility.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:442]: I don't see it as part of the PR diff, so I am assuming you mean it was already there somehow and it's correctly counting down.

File: server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java:192]: Why is this periodic? Seems like a one-shot. Maybe rename to `delayed`?

File: server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java:48]: `unpromotableRefreshRequestReceived`*

File: server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java:104]: Could add this assertion also to the test above?

File: server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java:110]: Do you mean with unpromotables?

**Review by kingherc — APPROVED** [review]
LGTM in general. Left some comments to take care of. I would recommend at least another approval before merging.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:3078]: What is the difference between this and simply returning `store.readLastCommittedSegmentsInfo().getLastGeneration()`?

File: server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:63]: One question: when/why `location` would be null? Should we also mark it with `@Nullabel`?

**Review by pxsalehi — COMMENTED** [review]
Is there a reason you haven't removed the work around in this PR? With the work-around we explicitly get rid of the refresh policy: `bulkRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.NONE)`.

> Also I think we should we remove the work-around in this PR: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java#L197-L199

Also could you please run the Serverless CI with a draft PR that updates the submodule with this (w/ removed workaround)? Please let me know if I have misunderstood the scope of the PR.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: The code has changed in this area.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Old code.

File: server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:63]: I marked it as nullable. Some transport write actions do not produce locations.

File: server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java
Comment (Tim-Brooks) [reply server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java:194]: Periodic refers to the type of refresh it is emulating. But I will probably rename it to scheduled refresh. See IndexShard#scheduledRefresh.

**Comment by Tim-Brooks** [comment]
@tlrx - I updated this with the approach we discuss. `FlushListeners` provides the logic. And it is implemented in `InternalEngine`. FYI, if this approach is reasonable then I think this PR is pretty much done minus the fact that I am going to add `FlushListeners` unit tests tomorrow.

I did already implement `InternalEngine` tests.

I also have a local PR for the stateless repo with a few minor changes and testing that refreshes still work. I will open that tomorrow.

File: server/src/main/java/org/elasticsearch/index/engine/FlushListeners.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/index/engine/FlushListeners.java:56]:
```suggestion
                throw new IllegalStateException("can't wait for flush on a closed index");
```

File: server/src/main/java/org/elasticsearch/index/engine/FlushListeners.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/index/engine/FlushListeners.java:77]: Maybe call this afterFlush?

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java:636]: Is it still needed?

File: server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:30]: do you think we could get rid of that nullable? I see it's only used in test but it makes non-testing code less easy to read

**Review by tlrx — COMMENTED** [review]
Looks good, I think this is the right way to implement this change. I made some comments but nothing to worry about. I'll get another final review but I expect it to be quick.

With this change we'll have to ensure to flush regularly if there are unpromotable shards. We could maybe rely on scheduled refreshes for this, or on flush-on-idle.

**Review by tlrx — APPROVED** [review]
LGTM, thanks for the iterations

**Pull request merged**
